From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Boy <sivertpaulsen2@gmail.com>
Date: Wed, 24 Jul 2024 13:53:58 +0200
Subject: [PATCH] Vault API

== AT ==
public net.minecraft.world.level.block.entity.vault.VaultServerData markChanged()V
public net.minecraft.world.level.block.entity.vault.VaultServerData getItemsToEject()Ljava/util/List;
public net.minecraft.world.level.block.entity.vault.VaultServerData getRewardedPlayers()Ljava/util/Set;
public net.minecraft.world.level.block.entity.vault.VaultServerData setItemsToEject(Ljava/util/List;)V

diff --git a/src/main/java/net/minecraft/world/level/block/entity/vault/VaultServerData.java b/src/main/java/net/minecraft/world/level/block/entity/vault/VaultServerData.java
index 85527fa85fa2d172ec7f142ce042c87bd2533029..4dd001d1ec9e178f348753fc08991455020a6b38 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/vault/VaultServerData.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/vault/VaultServerData.java
@@ -60,10 +60,19 @@ public class VaultServerData {
         return this.rewardedPlayers.contains(player.getUUID());
     }
 
+    // Paper start - Vault API
+    public boolean hasRewardedPlayer(UUID uniqueId) {
+        return this.rewardedPlayers.contains(uniqueId);
+    }
+
     @VisibleForTesting
     public void addToRewardedPlayers(Player player) {
-        this.rewardedPlayers.add(player.getUUID());
-        if (this.rewardedPlayers.size() > 128) {
+        addToRewardedPlayers(player.getUUID());
+    }
+
+    public void addToRewardedPlayers(UUID uniqueId) {
+        this.rewardedPlayers.add(uniqueId);
+        if (this.rewardedPlayers.size() > MAX_REWARD_PLAYERS) {
             Iterator<UUID> iterator = this.rewardedPlayers.iterator();
             if (iterator.hasNext()) {
                 iterator.next();
@@ -74,6 +83,12 @@ public class VaultServerData {
         this.markChanged();
     }
 
+    public void removeFromRewardedPlayers(UUID uniqueId) {
+        this.rewardedPlayers.remove(uniqueId);
+        this.markChanged();
+    }
+    // Paper end - Vault API
+
     long stateUpdatingResumesAt() {
         return this.stateUpdatingResumesAt;
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftVault.java b/src/main/java/org/bukkit/craftbukkit/block/CraftVault.java
index bfee498287c8358cbc8c9c2b7289d861cb90f0ad..1d437f6bc11dfb0505607d5ecf0cb31766177ef1 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftVault.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftVault.java
@@ -1,9 +1,25 @@
 package org.bukkit.craftbukkit.block;
 
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+import net.minecraft.resources.ResourceKey;
 import net.minecraft.world.level.block.entity.vault.VaultBlockEntity;
+import net.minecraft.world.level.block.entity.vault.VaultConfig;
+import org.bukkit.Bukkit;
 import org.bukkit.Location;
 import org.bukkit.World;
 import org.bukkit.block.Vault;
+import org.bukkit.craftbukkit.CraftLootTable;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.entity.Player;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.loot.LootTable;
+import org.jetbrains.annotations.Nullable;
 
 public class CraftVault extends CraftBlockEntityState<VaultBlockEntity> implements Vault {
 
@@ -15,6 +31,147 @@ public class CraftVault extends CraftBlockEntityState<VaultBlockEntity> implemen
         super(state, location);
     }
 
+    // Paper start - Vault API
+    @Override
+    public LootTable getLootTable() {
+        return CraftLootTable.minecraftToBukkit(getSnapshot().getConfig().lootTable());
+    }
+
+    @Override
+    public void setLootTable(final LootTable lootTable) {
+        ResourceKey<net.minecraft.world.level.storage.loot.LootTable> lootTableKey = CraftLootTable.bukkitToMinecraft(lootTable);
+        Preconditions.checkArgument(lootTableKey != null, "LootTable cannot be null");
+
+        VaultConfig config = this.getSnapshot().getConfig();
+        config = new VaultConfig(lootTableKey, config.activationRange(), config.deactivationRange(), config.keyItem(), config.overrideLootTableToDisplay());
+        this.getSnapshot().setConfig(config);
+    }
+
+    @Override
+    public @Nullable LootTable getOverrideDisplayLootTable() {
+        return getSnapshot().getConfig().overrideLootTableToDisplay().map(CraftLootTable::minecraftToBukkit).orElse(null);
+    }
+
+    @Override
+    public void setOverrideDisplayLootTable(@Nullable final LootTable lootTable) {
+        VaultConfig config = this.getSnapshot().getConfig();
+        config = new VaultConfig(config.lootTable(), config.activationRange(), config.deactivationRange(), config.keyItem(), Optional.ofNullable(CraftLootTable.bukkitToMinecraft(lootTable)));
+        this.getSnapshot().setConfig(config);
+    }
+
+    @Override
+    public ItemStack getKeyItem() {
+        return this.getSnapshot().getConfig().keyItem().asBukkitCopy();
+    }
+
+    @Override
+    public void setKeyItem(ItemStack keyItem) {
+        VaultConfig config = this.getSnapshot().getConfig();
+        config = new VaultConfig(config.lootTable(), config.activationRange(), config.deactivationRange(), CraftItemStack.asNMSCopy(keyItem), config.overrideLootTableToDisplay());
+        this.getSnapshot().setConfig(config);
+    }
+
+    @Override
+    public double getActivationRange() {
+        return this.getSnapshot().getConfig().activationRange();
+    }
+
+    @Override
+    public void setActivationRange(double activationRange) {
+        VaultConfig config = this.getSnapshot().getConfig();
+        Preconditions.checkArgument(activationRange <= config.activationRange(), "ActivationRange must be greater or equal to DeactivationRange");
+
+        config = new VaultConfig(config.lootTable(), activationRange, config.deactivationRange(), config.keyItem(), config.overrideLootTableToDisplay());
+        this.getSnapshot().setConfig(config);
+    }
+
+    @Override
+    public double getDeactivationRange() {
+        return this.getSnapshot().getConfig().deactivationRange();
+    }
+
+    @Override
+    public void setDeactivationRange(final double deactivationRange) {
+        VaultConfig config = this.getSnapshot().getConfig();
+        Preconditions.checkArgument(deactivationRange >= config.activationRange(), "DeactivationRange must be greater or equal to ActivationRange");
+
+        config = new VaultConfig(config.lootTable(), config.activationRange(), deactivationRange, config.keyItem(), config.overrideLootTableToDisplay());
+        this.getSnapshot().setConfig(config);
+    }
+
+    @Override
+    public Set<Player> getRewardedPlayers() {
+        ImmutableSet.Builder<Player> players = ImmutableSet.builder();
+
+        for (UUID uuid : getRewardedPlayerUUIDs()) {
+            Player player = Bukkit.getPlayer(uuid);
+            if (player != null) {
+                players.add(player);
+            }
+        }
+        return players.build();
+    }
+
+    @Override
+    public Set<UUID> getRewardedPlayerUUIDs() {
+        return ImmutableSet.copyOf(this.getTileEntity().getServerData().getRewardedPlayers());
+    }
+
+    @Override
+    public boolean hasRewardedPlayer(final Player player) {
+        Preconditions.checkArgument(player != null, "Player cannot be null");
+
+        return hasRewardedPlayer(player.getUniqueId());
+    }
+
+    @Override
+    public boolean hasRewardedPlayer(final UUID uniqueId) {
+        Preconditions.checkArgument(uniqueId != null, "UUID cannot be null");
+
+        return this.getTileEntity().getServerData().hasRewardedPlayer(uniqueId);
+    }
+
+    @Override
+    public void addToRewardedPlayers(Player player) {
+        Preconditions.checkArgument(player != null, "Player cannot be null");
+
+        addToRewardedPlayers(player.getUniqueId());
+    }
+
+    @Override
+    public void addToRewardedPlayers(final UUID uniqueId) {
+        Preconditions.checkArgument(uniqueId != null, "UUID cannot be null");
+
+        this.getTileEntity().getServerData().addToRewardedPlayers(uniqueId);
+    }
+
+    @Override
+    public void removeFromRewardedPlayers(Player player) {
+        Preconditions.checkArgument(player != null, "Player cannot be null");
+
+        removeFromRewardedPlayers(player.getUniqueId());
+    }
+
+    @Override
+    public void removeFromRewardedPlayers(final UUID uniqueId) {
+        Preconditions.checkArgument(uniqueId != null, "UUID cannot be null");
+
+        this.getTileEntity().getServerData().removeFromRewardedPlayers(uniqueId);
+    }
+
+    @Override
+    public List<ItemStack> getItemsToEject() {
+        return this.getTileEntity().getServerData().getItemsToEject().stream().map(net.minecraft.world.item.ItemStack::asBukkitCopy).toList();
+    }
+
+    @Override
+    public void setItemsToEject(final List<ItemStack> itemsToEject) {
+        Preconditions.checkArgument(itemsToEject != null, "ItemsToEject cannot be null");
+
+        this.getTileEntity().getServerData().setItemsToEject(itemsToEject.stream().map(CraftItemStack::asNMSCopy).toList());
+    }
+    // Paper end - Vault API
+
     @Override
     public CraftVault copy() {
         return new CraftVault(this, null);
diff --git a/src/main/java/org/bukkit/craftbukkit/block/data/type/CraftVault.java b/src/main/java/org/bukkit/craftbukkit/block/data/type/CraftVault.java
index 5a5f60ba2c6b6ad29693daf75eaab83ac6c1b816..808d3ae5270e5bfe73dd163fd1150b0ab470f643 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/data/type/CraftVault.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/data/type/CraftVault.java
@@ -9,12 +9,12 @@ public abstract class CraftVault extends CraftBlockData implements Vault {
     private static final net.minecraft.world.level.block.state.properties.BooleanProperty OMINOUS = getBoolean("ominous");
 
     @Override
-    public org.bukkit.block.data.type.Vault.State getTrialSpawnerState() {
+    public org.bukkit.block.data.type.Vault.State getVaultState() { // Paper
         return this.get(CraftVault.VAULT_STATE, org.bukkit.block.data.type.Vault.State.class);
     }
 
     @Override
-    public void setTrialSpawnerState(org.bukkit.block.data.type.Vault.State state) {
+    public void setVaultState(org.bukkit.block.data.type.Vault.State state) { // Paper
         this.set(CraftVault.VAULT_STATE, state);
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/block/impl/CraftVault.java b/src/main/java/org/bukkit/craftbukkit/block/impl/CraftVault.java
index 4c62caba7dff6538309cac46861fed047223bb5f..3db5ca6fe3d54cfd6285ee57bd242e72a2ec0c3a 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/impl/CraftVault.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/impl/CraftVault.java
@@ -19,12 +19,12 @@ public final class CraftVault extends org.bukkit.craftbukkit.block.data.CraftBlo
     private static final net.minecraft.world.level.block.state.properties.BooleanProperty OMINOUS = getBoolean(net.minecraft.world.level.block.VaultBlock.class, "ominous");
 
     @Override
-    public org.bukkit.block.data.type.Vault.State getTrialSpawnerState() {
+    public org.bukkit.block.data.type.Vault.State getVaultState() { // Paper
         return this.get(CraftVault.VAULT_STATE, org.bukkit.block.data.type.Vault.State.class);
     }
 
     @Override
-    public void setTrialSpawnerState(org.bukkit.block.data.type.Vault.State state) {
+    public void setVaultState(org.bukkit.block.data.type.Vault.State state) { // Paper
         this.set(CraftVault.VAULT_STATE, state);
     }
 
